Update jackson version and add constraints for jackson core

irfano opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This resolves depandabot security issue: https://github.com/woocommerce/woocommerce-android/security/dependabot/152

The warning suggests that we update `com.fasterxml.jackson.core:jackson-core` to version 2.15.0.

The Jackson library is used by the WireMock library. Since we can’t update WireMock, we set constraints for the libraries WireMock uses. I’ve added an additional constraint for the Jackson core. I also updated the version of the jackson databinding library, this update wasn’t required for the warning itself, but it’s better to keep all Jetty libraries on the same version.

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
- Green check for UI tests.

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

irfano committed:
Add jackson core library

This was added to set constraints for other libraries.

irfano committed:
Use common jackson version for jackson databind

irfano committed:
Merge branch 'trunk' into add-jackson-core-library

irfano committed:
Merge branch 'trunk' into add-jackson-core-library

hichamboushaba commented on gradle/libs.versions.toml:
```diff
@@ -78,7 +78,7 @@
 google-zxing = '3.5.3'
 gravatar = '0.2.0'
-jackson-databind = '2.12.7.1'
+jackson = '2.15.0'
```
@irfano can't we use the newer version here?

hichamboushaba replied:
Actually, why not use the last version `[2.19.2](https://mvnrepository.com/artifact/com.fasterxml.jackson.core/jackson-core/2.19.2)`? Jirka is having a separate PR to update `jackson-databind` to the last version, so we could probably unify the efforts.

irfano replied:
I made some changes after your comment but then reverted them. I also considered setting the latest version of the Jackson library, but here are the reasons why I chose the minimum safe version instead:
This library is only used transitively by the WireMock library. By adding constraints, we’re forcing WireMock to use a higher version of Jackson than the one it was originally built with. Upgrading Jackson too far introduces a risk of incompatibility between WireMock and Jackson. That’s why I chose version 2.15.0.

I’ll wait for your response before merging, @hichamboushaba.

malinajirka replied:
I understand the reasoning and I can't say I'd fully disagree. Having said that there are two reasons I'm inclined towards updating to the newest version.

1. If we upgrade to 2.15.0, there is a chance dependabot will immediatelly create another PR request for updating to 2.19.2. 
2. Moreover, the fact that this specific library is used only by WireMock is just a coincidence - we have tens of other libraries that are used by mutliple dependencies and we are force-updating them. So there IMO isn't much reason to make an exception here just because it is currently used only by one library. 
 
Wdyt?

irfano replied:
Thanks for sharing your thoughts, @malinajirka!

> 1. If we upgrade to 2.15.0, there is a chance dependabot will immediatelly create another PR request for updating to 2.19.2.

Actually it won't, we are ignoring Jackson Core library upgrades in the [dependabot.yml](https://github.com/woocommerce/woocommerce-android/blob/c485cb2ef3a402ef98360c8d9fbc98428bd538df/.github/dependabot.yml#L32-L33) file. 

> 2. Moreover, the fact that this specific library is used only by WireMock is just a coincidence - we have tens of other libraries that are used by mutliple dependencies and we are force-updating them. So there IMO isn't much reason to make an exception here just because it is currently used only by one library.

There are also other dependencies used only transitively that we force-upgrade. You can see we're ignoring those as well in [dependabot.yml](https://github.com/woocommerce/woocommerce-android/blob/c485cb2ef3a402ef98360c8d9fbc98428bd538df/.github/dependabot.yml#L19-L35). For example, I followed the same approach for the [Jetty Webapp upgrade](https://github.com/woocommerce/woocommerce-android/pull/14409).

malinajirka replied:
Thanks for providing the additional details! 

> There are also other dependencies used only transitively that we force-upgrade solely for security vulnerabilities. You can see we're ignoring those as well in dependabot.yml. For example, I followed the same approach for the Jetty Webapp upgrade.

Just to clarify, I'm not against force-updating, I think that's completely valid. What I wanted to say was that IMO it's completely fine to update a dependency to a newer version than the one that is referenced by the library, since we do this in hundreds of other cases anyway => in other words `we’re forcing WireMock to use a higher version of Jackson than the one it was originally built with.` while it is true, I think it shouldn't concern us.

Having said all that, final decision is up to you, I'm completely fine with both. I'm not sure what is better tbh.

irfano replied:
So, if we use the latest versions of Wiremock's transitive dependencies and try to keep them up to date, we may run into issues in UI tests due to inconsistencies between WireMock 2.26.3 and its updated dependencies. These incompatibilities might be subtle and hard to detect, especially in UI tests.

So, unless there’s a critical reason to keep these transitive dependencies up to date, I prefer sticking to the minimum safe versions.

irfano replied:
> Having said all that, final decision is up to you, I'm completely fine with both.

Thank you both for the discussion. I’m proceeding with the merge.

hichamboushaba reviewed (approved):
LGTM, I just left a question about the newer version that's available.

irfano committed:
Upgrade Jackson to 2.19.2

irfano committed:
Update ignored jackson dependency in Dependabot config

The `jackson-databind` dependency is replaced with `jackson-core` in the list of ignored dependencies for Dependabot.

irfano committed:
Merge branch 'trunk' into add-jackson-core-library

irfano committed:
Revert "Upgrade Jackson to 2.19.2"

This reverts commit 3647780826213e37fc719ac50ec1cc3197f43096.

irfano committed:
Revert "Update ignored jackson dependency in Dependabot config"

This reverts commit 27a83e98d5e065cb9b369527ac99c96a187f49ee.

irfano committed:
Update ignored jackson dependency in Dependabot config

The `jackson-core` dependency is added to the list of ignored dependencies for Dependabot.
